pkg/agent/multicluster/mc_route_controller.go

```diff
@@ -0,0 +1,460 @@
+	}
+
+	if activeGW != nil {
+		if activeGW.Name == c.nodeConfig.Name {
+			if err := c.ofClient.InstallMulticlusterClassifierFlows(activeGW.Name, config.DefaultTunOFPort, true); err != nil {
+				return err
+			}
+			c.installedActiveGWName = activeGW.Name
+			return c.addMCFlowsForAllCIImps(activeGW)
+		} else {
+			if err := c.ofClient.InstallMulticlusterClassifierFlows(activeGW.Name, config.DefaultTunOFPort, false); err != nil {
+				return err
+			}
+			c.installedActiveGWName = activeGW.Name
+			return c.addMCFlowsForAllCIImps(activeGW)
+		}
+	}
+	return nil
+}
+
+func (c *MCRouteController) syncMCFlowsForAllCIImps(activeGW *mcv1alpha1.Gateway) error {
+	installedCIImportsCache := c.installedCIImports.List()
+	desiredCIImports, err := c.ciImportLister.ClusterInfoImports(c.namespace).List(labels.Everything())
+	if err != nil {
+		return err
+	}
+	installedCIImports := map[string]*mcv1alpha1.ClusterInfoImport{}
+	for _, ciImp := range installedCIImportsCache {
+		ciImport := ciImp.(*mcv1alpha1.ClusterInfoImport)
+		installedCIImports[ciImport.Name] = ciImport
+	}
+
+	for ciImpName, ciImp := range installedCIImports {
+		if !c.containsCIImport(desiredCIImports, ciImpName) {
+			if err := c.deleteMCFlowsForCIImport(ciImp); err != nil {
+				return err
+			}
+		}
+	}
+
+	for _, ciImp := range desiredCIImports {
+		if err = c.addMCFlowsForSingleCIImp(activeGW, ciImp, installedCIImports[ciImp.Name]); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func (c *MCRouteController) containsCIImport(ciImports []*mcv1alpha1.ClusterInfoImport, ciImportName string) bool {
+	for _, cii := range ciImports {
+		if cii.Name == ciImportName {
+			return true
+		}
+	}
+	return false
+}
+
+func (c *MCRouteController) cleanupFlows(gatewayName string) error {
+	if err := c.deleteMCFlowsForAllCIImps(); err != nil {
+		return err
+	}
+	if err := c.ofClient.UninstallMulticlusterFlows(gatewayName); err != nil {
+		return err
+	}
+	klog.V(2).InfoS("Deleted flows for Gateway", "gateway", gatewayName)
+	return nil
+}
+
+func (c *MCRouteController) addMCFlowsForAllCIImps(activeGW *mcv1alpha1.Gateway) error {
+	allCIImports, err := c.ciImportLister.ClusterInfoImports(c.namespace).List(labels.Everything())
+	if err != nil {
+		return err
+	}
+	if len(allCIImports) == 0 {
+		klog.V(2).InfoS("No remote ClusterInfo imported, do nothing")
+		return nil
+	}
+
+	for _, ciImport := range allCIImports {
+		if err := c.addMCFlowsForSingleCIImp(activeGW, ciImport, nil); err != nil {
+			if strings.Contains(err.Error(), "invalid Gateway IP") {
+				continue
+			}
+			return err
+		}
+	}
+
+	return nil
+}
+
+func (c *MCRouteController) addMCFlowsForSingleCIImp(activeGW *mcv1alpha1.Gateway, ciImport *mcv1alpha1.ClusterInfoImport, installedCIImp *mcv1alpha1.ClusterInfoImport) error {
+	tunnelPeerIP := getPeerGatewayIP(ciImport.Spec)
+	if tunnelPeerIP == nil {
+		return errors.New("invalid Gateway IP")
+	}
+	if installedCIImp != nil && reflect.DeepEqual(ciImport, installedCIImp) {
```

Comment:
Can we just compare the three IPs, rather than using DeepEqual?

Reply:
Done